### PR TITLE
CDRO-2354 [windows] Role permissions for older ElasticSearch releases

### PR DESCRIPTION
**Pull request by Muxamedin**

### CODE REVIEW

**Review by urvashisingh — APPROVED** [review]

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 1]
```diff
@@ -0,0 +1,528 @@
+<#
+.SYNOPSIS
+This script must change roles permissions for user of ElasticSearch installed
+by CloudBees Software Delivery Automation Analytics.
+Script require Admin permissions.
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:5]:
```suggestion
This script updates ElasticSearch user permissions for CloudBees Analytics in v2023.10.0 or earlier. If you are using v2023.12.0 or later, this update is not required.

IMPORTANT:  This script requires Administrator permissions to run.
```

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+
+.DESCRIPTION
+CloudBees Software Delivery Automation Analytics moved from ElasticSearch to OpenSearch in a release 2024.06
+To migrate from ElasticSearch to OpenSearch starting 2023.12 and later no need any changes. If any issues
+with older release in time of migration - current script can help with role permissions issues.
+Aplay this script before migrate from 2023.8 or older to the 2024.06 or younger.
+Script will try automatically find all needed values and if any issues to do it  - will report you about issues.
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:12]:
```suggestion
CloudBees Analytics migrated from ElasticSearch to OpenSearch in release v2024.06.0. This script is required update user permissons when upgrading from v2023.10 or earlier as part of the migration when upgrading to v2024.06.0 or later. 
This script attempts to automatically find all required values silently. However, if an issue is encountered, it reports the issue as an error messages.
```

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+
+.PARAMETER Port
+Specify CloudBees Software Delivery Automation Analytics transport port if you know it. Script will try to read port namber from config file.
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:15]:
```suggestion
.PARAMETER Port
Specify the CloudBees Analytics transport port. The script tries to read port number automatically from the config file.
```

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+
+
+.PARAMETER Data
+Specify path to the Data folder if it is not installed by default or script was not able figureout location of that folder.
+By default it will be installed as C:\ProgramData\CloudBees\Software Delivery Automation\ .
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:20]:
```suggestion
Specify path to the Data directory, if it is not installed in the default location, or script was not able to discover the data directory.
By default, the Data directory is C:\ProgramData\CloudBees\Software Delivery Automation\ .
```

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 5]
```diff
@@ -0,0 +1,528 @@
...
+
+.PARAMETER Install
+Specify path to the installation directory.
+By default it will be installed as C:\Program Files\CloudBees\Software Delivery Automation\ .
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:24]:
```suggestion
Specify path to the installation directory.
By default, the Install directory is C:\Program Files\CloudBees\Software Delivery Automation\ .
```

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 6]
```diff
@@ -0,0 +1,528 @@
...
+
+.PARAMETER Temp
+Path to the temporary folder with write permissions. If parameter will not be specifyed script will use directory $env:TEMP
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:27]:
```suggestion
Path to a temporary directory with write permissions required as part of the migration. If the Temp parameter is not specified, the script uses $env:TEMP.
```

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 7]
```diff
@@ -0,0 +1,528 @@
...
+
+.PARAMETER JavaHome
+Path to the Java Home directory. By default will be used java from install directory.
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:30]:
```suggestion
Path to the Java Home directory. By default, the Java included in the CloudBees Analytics install directory is used.
```

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 8]
```diff
@@ -0,0 +1,528 @@
...
+
+.PARAMETER Help
+Script will print usage and quit.
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:33]:
```suggestion
Prints usage and quits.
```

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 9]
```diff
@@ -0,0 +1,528 @@
...
+
+
+.EXAMPLE
+# if application installed to the  "C:\Program Files\DOIS\" and config installed to the  "C:\ProgramData\DOIS\" and transport port is 9301
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:37]:
```suggestion
# The CloudBees Analytics installation directory is "C:\Program Files\DOIS\", the data directory is "C:\ProgramData\DOIS\", and the transport port is 9301:
```

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 10]
```diff
@@ -0,0 +1,528 @@
...
+PS> update-role-permissions.ps1 -Install "C:\Program Files\DOIS\CloudBees\Software Delivery Automation" -Data "C:\ProgramData\DOIS\CloudBees\Software Delivery Automation" -Port 9301
+
+.EXAMPLE
+# if port 9301
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:41]:
```suggestion
# The transport port is 9301:
```

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 11]
```diff
@@ -0,0 +1,528 @@
...
+PS> update-role-permissions.ps1 -Port 9301
+# or
+PS> update-role-permissions.ps1 -p 9301
+
+.EXAMPLE
+# to see usage
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:47]:
```suggestion
# To view the script usage:
```

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 12]
```diff
@@ -0,0 +1,528 @@
...
+    [string]$Port,
+    [string]$Data,
+    [string]$Install,
+    [string]$Temp,
+    [string]$JavaHome
+)
+
+
+# Setup values to manage script
+# Apllication name to look for information about it at the Windows registry
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:64]:
```suggestion
# Application name used to look for information in the the Windows registry:
```

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 13]
```diff
@@ -0,0 +1,528 @@
...
+$LOG_TO_FILE = $true  # false if you dont need to log to file
+$DEBUG = $false  # $true - to write debug messages, $false - debug messages will not appear in a log file
+
+function Show-Usage {
+    $usage = @"
+Usage: update-role-permissions.ps1 [-Help] [-Port <number>] [-Data <path>] [-Install <path>] [-Temp <path>] [-JavaHome <path>]
+
+Parameters:
+    -Help          Show usage.
+    -Port          Specify trasport prot for CloudBees Software Delivery Automation Analytics.
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:76]:
```suggestion
    -Port          Specify the trasport port for CloudBees Analytics.
```

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 14]
```diff
@@ -0,0 +1,528 @@
...
+    -Data          Data directory, contains config and yaml files
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:77]:
```suggestion
    -Data          Specify the Data directory that contains the CloudBees Analytics config and YAML files.
```

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 15]
```diff
@@ -0,0 +1,528 @@
...
+    -Install       Path to the installation folder
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:78]:
```suggestion
    -Install        Specify the Path to the CloudBees Analytics installation folder.
```

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 16]
```diff
@@ -0,0 +1,528 @@
...
+    -Temp          Temporary directory with write permissions, will be used for temporary config files
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:79]:
```suggestion
    -Temp          Temporary directory with write permissions used for temporary config files while updating. 
```

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 17]
```diff
@@ -0,0 +1,528 @@
...
+    -JavaHome      Path to the JAVA HOME you like to use
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:80]:
```suggestion
    -JavaHome      Path to the JAVA HOME to use. 
```

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 18]
```diff
@@ -0,0 +1,528 @@
...
+if (-not (Test-Administrator) -and (-not $Analyze) ) {
+    Write-Host "Please run the script with Administrator permissions. Quit." -ForegroundColor Yellow
+    exit
+}
+
+# ES_DATA - data directory, contains config and yaml files
+# ES_INSTALL - installation  direrctory for DOIS with ElasticSearch
+# ES_JAVA_HOME - Home for java
+# ES_PORT - the transport port used for communication between nodes, specify new port if it was not configured by default as 9300
+# CONFIG_DIR_TMP - temporary folder with write permissions
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:289]:
```suggestion
# ES_DATA - Data directory, contains config and YAML files
# ES_INSTALL - Installation  direrctory for DOIS with ElasticSearch
# ES_JAVA_HOME - Java Home path
# ES_PORT - The transport port used for communication between nodes, specify a new port if it was not configured to the default 9300
# CONFIG_DIR_TMP - Temporary folder with write permissions used during migration
```

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 19]
```diff
@@ -0,0 +1,528 @@
...
+    Write-Log -level "info" -message "Passed Port from command line as: $Port"
+    $ES_PORT = $Port
+}
+
+if ($Data) {
+    Write-Log -level "info" -message "Passed path to Data folder from command line as: $Data"
+    $ES_DATA = $Data
+}
+if ($Install) {
+    Write-Log -level "info" -message "Passed installation directory from from command line as: $Install"
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:309]:
```suggestion
    Write-Log -level "info" -message "Passed Install directory from from command line as: $Install"
```

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 20]
```diff
@@ -0,0 +1,528 @@
...
+    $ES_INSTALL = $Install
+}
+
+if ($Temp) {
+    Write-Log -level "info" -message "Passed path to temporary directory with write permissions from command line as: $Temp"
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:314]:
```suggestion
    Write-Log -level "info" -message "Passed path to Temp directory with write permissions from command line as: $Temp"
```

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 21]
```diff
@@ -0,0 +1,528 @@
...
+    } catch {
+        $knowPortNumber = $false
+    }
+
+    if ($portNumber) {
+        Write-Log -level "info" -message "transport port number : $portNumber"
+    }
+
+    if ( -not $knowPortNumber ) {
+        Write-Log -level "error" -message "There is no reason to continue as not all parameters was found. Quit."
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:498]:
```suggestion
        Write-Log -level "error" -message "Can not continue as not all parameters were found. Quit."
```

File: cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1 [hunk 22]
```diff
@@ -0,0 +1,528 @@
...
+
+$args =  "-cp", "$searchGuard7Lib\*;$elasticSearchLib\*",  `
+  "com.floragunn.searchguard.tools.SearchGuardAdmin", `
+  "-cd", "$temporaryConfigDirectory", `
+  "-ks", "$adminkeystore", "-kspass", "abcdef", `
+  "-ts", "$truststore", "-tspass", "abcdef", `
+  "-h", "localhost", "-p", $portNumber, "-nhnv", "-icl"
+
+
+Write-Log -level "debug" -message "Goign to execute next command: & $javaBin $args"
```
Comment (jsmith-cb) [inline cloudbees-cd/es-to-os-migration/update-es-role-permission.ps1:512]:
```suggestion
Write-Log -level "debug" -message "Executing next command: & $javaBin $args"
```

**Review by Muxamedin — COMMENTED** [review]
accepted proposed changes

**Review by jsmith-cb — APPROVED** [review]

**Pull request merged**